userspace/libpman/src/capture.c

```diff
@@ -19,53 +19,115 @@ limitations under the License.
 #include <libpman.h> // before including scap so that handle_ppm_sc_mask() is not built
 #include <scap.h>
 
-int pman_enable_capture(bool *sc_set)
+/* This function should be idempotent, every time it is called it should enforce again the state */
+int pman_enforce_sc_set(bool *sc_set)
 {
-	if (!sc_set)
+	/* If we fail at initialization time the BPF skeleton
+	 * is not initialized when we stop the capture for example
+	 */
+	if(!g_state.skel)
 	{
-		for (int i = 0; i < PPM_SC_MAX; i++)
-		{
-			pman_mark_single_ppm_sc(i, true);
-		}
-		return pman_attach_all_programs();
+		return SCAP_FAILURE;
+	}
+
+	/* When we want to disable the capture we receive a NULL pointer here */
+	bool empty_sc_set[PPM_SC_MAX] = {0};
+	if(!sc_set)
+	{
+		sc_set = empty_sc_set;
 	}
 
 	int ret = 0;
-	/* Enable requested tracepoints */
-	bool tp_set[TP_VAL_MAX];
-	tp_set_from_sc_set(sc_set, tp_set);
-	for (int i = 0; i < TP_VAL_MAX && ret == 0; i++)
+	int syscall_id = 0;
+	/* Special tracepoints, their attachment depends on interesting syscalls */
+	bool sys_enter = false;
+	bool sys_exit = false;
+	bool sched_prog_fork = false;
+	bool sched_prog_exec = false;
+
+	/* Enforce interesting syscalls */
+	for(int sc = 0; sc < PPM_SC_MAX; sc++)
 	{
-		if(tp_set[i])
+		syscall_id = scap_ppm_sc_to_native_id(sc);
+		/* if `syscall_id` is -1 this is not a syscall */
+		if(syscall_id == -1)
 		{
-			ret = pman_update_single_program(i, true);
+			continue;
```

Comment:
Since these SCs will be the ones attached to tracepoints, can we add a `switch` statement here instead of the `if/else` logic below?
We could even add a `default` case with an `ASSERT(false)` so that one would immediately realize if she/he forgot to manage the new tp!

Reply:
unfortunately no, because `scap_ppm_sc_to_native_id()` uses the syscall_table to retrieve the syscall id, so there are some PPM_SC that are syscalls but they cannot be associated with the right syscall id because the mapping into syscall_table is compiled out with `ifdefs`

Reply:
:/ well, i guess it's ok then!